### PR TITLE
Added pod evictors for new zone.

### LINKED ISSUE #47791: Adding `failure-domain.beta.kubernetes.io/zone` label manually causes node controller panic
**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.1.51+594edb77397512-dirty", GitCommit:"594edb773975124b137fa12bc573478a0355e89c", GitTreeState:"dirty", BuildDate:"2017-06-20T10:51:10Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.1.51+594edb77397512-dirty", GitCommit:"594edb773975124b137fa12bc573478a0355e89c", GitTreeState:"dirty", BuildDate:"2017-06-20T11:29:14Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}

**What happened**: After adding `failure-domain.beta.kubernetes.io/zone` label to a node, node controller panics.

**What you expected to happen**: Everything should still work.

**How to reproduce it** (as minimally and precisely as possible):

Run `"cluster/kubectl.sh" label  nodes --all failure-domain.beta.kubernetes.io/zone=zone` and wait for the node controller to panic:

```
E0620 14:45:34.888655       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/home/salmon/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:514
/home/salmon/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:489
/home/salmon/.gimme/versions/go1.8.linux.amd64/src/runtime/panic.go:63
/home/salmon/.gimme/versions/go1.8.linux.amd64/src/runtime/signal_unix.go:290
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/rate_limited_queue.go:244
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:1023
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:659
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:490
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98
/home/salmon/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52
/home/salmon/.gimme/versions/go1.8.linux.amd64/src/runtime/asm_amd64.s:2197
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3484942]
```

**Anything else we need to know**:

I've checked the code and it looks like the node controller expects the label to be present during node registration and it can not be added later.
Even if the label can not be added latter it shouldn't panic.

### PR DESCRIPTION
**Pull request by k82cn**
**What this PR does / why we need it**:
Added pod evictors for new zone which was added manually.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #47791 

**Release note**:
```release-note-none
```

### CODE REVIEW

**Comment by k82cn** [comment]
/unassign @aveshagarwal @gmarek

I'll add UT for it; and will re-assign to you when it's ready to review.

**Comment by k82cn** [comment]
/unassign  @ingvagabund @smarterclayton

sorry for disruption :(.

**Comment by k82cn** [comment]
/assign @aveshagarwal @gmarek

Here's the description in the doc, so I think we need to handle this issue :).

>Example:
>
>failure-domain.beta.kubernetes.io/region=us-east-1
>
>failure-domain.beta.kubernetes.io/zone=us-east-1c
>
>Used on: Node, PersistentVolume
>
>On the Node: Kubelet populates this with the zone information as defined by the cloudprovider. It will not be set if not using a cloudprovider, but you should consider setting it on the nodes if it makes sense in your topology.

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -991,12 +1000,19 @@ func (nc *NodeController) tryUpdateNodeStatus(node *v1.Node) (time.Duration, v1.
 	return gracePeriod, observedReadyCondition, currentReadyCondition, err
 }
 
-func (nc *NodeController) checkForNodeAddedDeleted(nodes []*v1.Node) (added, deleted []*v1.Node) {
+func (nc *NodeController) checkForNodeAddedDeleted(nodes []*v1.Node) (added, deleted, updated []*v1.Node) {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I don't like wording. 'updated' accompanied by 'added' and 'deleted' suggests that it means updated Nodes, not new zones. Please rename it to sth like 'newZoneRepresentatives' and add a comment explaining return values.

Comment (k82cn) [inline pkg/controller/node/nodecontroller.go]: done, I renamed it to "classifyNodes" which seems better :).

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -991,21 +1000,32 @@ func (nc *NodeController) tryUpdateNodeStatus(node *v1.Node) (time.Duration, v1.
...
 
-func (nc *NodeController) checkForNodeAddedDeleted(nodes []*v1.Node) (added, deleted []*v1.Node) {
-	for i := range nodes {
-		if _, has := nc.knownNodeSet[nodes[i].Name]; !has {
-			added = append(added, nodes[i])
+// classifyNodes classifies the allNodes to three categories:
+//   1. added: the nodes that in 'allNodes', but not in 'knownNodeSet'
+//   2. deleted: the nodes that in 'knownNodeSet', but not in 'allNodes'
+//   3. updated: the nodes that in both 'knownNodeSet' and 'allNodes', but no zone states
+func (nc *NodeController) classifyNodes(allNodes []*v1.Node) (added, deleted, updated []*v1.Node) {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I meant that I don't like 'updated' name of the return variable:)

Comment (k82cn) [inline pkg/controller/node/nodecontroller.go]: updated to `newZoneRepresentatives` :).

**Pull request merged**